graph/test_wgraph.py

```diff
@@ -0,0 +1,206 @@
+    for node in temp_nodes:
+        empty_graph.add_node(node)
+        if prev is not None:
+            empty_graph.add_edge(prev, node, 1)
+        prev = node
+    return empty_graph
+
+
+@pytest.fixture()
+def circle_graph(temp_nodes, linear_graph):
+    linear_graph.add_edge(temp_nodes[3], temp_nodes[0], 1)
+    return linear_graph
+
+
+@pytest.fixture()
+def loop_graph(temp_nodes, linear_graph):
+    linear_graph.add_edge(temp_nodes[0], temp_nodes[3], 1)
+    return linear_graph
+
+
+@pytest.fixture()
+def branch_graph_and_nodes(empty_graph):
+    na = Node('a')
+    nb = Node('b')
+    nc = Node('c')
+    nd = Node('d')
+    ne = Node('e')
+    nf = Node('f')
+    ng = Node('g')
+    empty_graph.add_edge(na, nb, 1)
+    empty_graph.add_edge(na, nc, 1)
+    empty_graph.add_edge(nb, nd, 1)
+    empty_graph.add_edge(nb, ne, 1)
+    empty_graph.add_edge(nb, nf, 1)
+    empty_graph.add_edge(nc, nf, 1)
+    empty_graph.add_edge(nd, ng, 1)
+    return empty_graph, [na, nb, nc, nd, ne, nf, ng]
+
+
+def test_constructor(empty_graph):
+    assert isinstance(empty_graph, Graph)
+
+
+def test_hasnode(empty_graph):
+    n1 = Node('a')
+    empty_graph._data = {n1: set()}
+    assert empty_graph.has_node(n1)
+
+
+def test_addnode(empty_graph):
+    n1 = Node('a')
+    n2 = Node('b')
+    n3 = Node(123)
+    empty_graph.add_node(n1)
+    empty_graph.add_node(n2)
+    empty_graph.add_node(n3)
+    assert empty_graph.has_node(n1)
+    assert empty_graph.has_node(n2)
+    assert not empty_graph.has_node(123)
+    assert empty_graph.has_node(n3)
+    list_nodes = empty_graph.nodes()
+    assert n1 in list_nodes
+    assert n2 in list_nodes
+    assert n3 in list_nodes
+
+
+def test_delnode(graph_nodes, temp_nodes):
+    assert graph_nodes.has_node(temp_nodes[0]) is True
+    graph_nodes.del_node(temp_nodes[0])
+    assert graph_nodes.has_node(temp_nodes[0]) is False
+    with pytest.raises(KeyError):
+        graph_nodes.del_node(temp_nodes[0])
+
+
+def test_edges(graph_nodes, temp_nodes):
+    assert len(graph_nodes.edges()) == 4
+    graph_nodes.add_edge(temp_nodes[0], temp_nodes[1], 1)
+    assert (temp_nodes[0], {temp_nodes[1]: 1}) in graph_nodes.edges()
+    assert (temp_nodes[1], {temp_nodes[0]: 1}) not in graph_nodes.edges()
+    graph_nodes.add_edge(temp_nodes[2], temp_nodes[1], 5)
+    assert (temp_nodes[2], {temp_nodes[1]: 5}) in graph_nodes.edges()
+    graph_nodes.add_edge(temp_nodes[1], temp_nodes[2], 2)
+    assert (temp_nodes[1], {temp_nodes[2]: 2}) in graph_nodes.edges()
+    graph_nodes.add_edge(temp_nodes[1], 'b', 1)
+    assert graph_nodes.has_node('b')
+    assert ('b', {}) in graph_nodes.edges()
+    graph_nodes.del_edge(temp_nodes[1], 'b')
+    assert graph_nodes.has_node('b') is True
+    assert (temp_nodes[1], 'b') not in graph_nodes.edges()
+    with pytest.raises(KeyError):
+        graph_nodes.del_edge(temp_nodes[1], 'b')
+    with pytest.raises(KeyError):
+        graph_nodes.del_edge(temp_nodes[1], 'imaginarynode')
```

Comment:
This is a pretty hairy test.. Can you find a way to make it more simple or cleaned up?